### ISSUE TITLE
Using Undirected Graphs in routing.route_to_gdf

### ISSUE THREAD

**Issue by aarongraybill** [gboeing/osmnx#1257]
### Contributing guidelines

- [X] I understand the contributing guidelines

### Documentation

- [X] My problem is not addressed by the documentation or examples

### Existing issues

- [X] My problem does not appear in an existing issue

### What operating system and Python version are you using?

macOS Sonoma, python 3.12.3

### What OSMnx version are you using?

2.0.0

### Environment packages and versions

```shell
attrs==24.2.0
branca==0.8.0
Cartopy==0.23.0
certifi==2024.8.30
charset-normalizer==3.3.2
click==8.1.7
click-plugins==1.1.1
cligj==0.7.2
contourpy==1.3.0
cycler==0.12.1
fiona==1.10.0
folium==0.17.0
fonttools==4.53.1
geopandas==1.0.1
idna==3.8
Jinja2==3.1.4
joblib==1.4.2
kiwisolver==1.4.7
mapclassify==2.8.1
MarkupSafe==2.1.5
matplotlib==3.9.2
networkx==3.3
numpy==1.26.4
osmnx==2.0.0
packaging==24.1
pandas==2.2.2
pillow==10.4.0
pyogrio==0.10.0
pyparsing==3.1.4
pyproj==3.6.1
pyshp==2.3.1
python-dateutil==2.9.0.post0
pytz==2024.1
requests==2.32.3
scikit-learn==1.5.1
scipy==1.14.1
setuptools==74.1.2
shapely==2.0.6
six==1.16.0
threadpoolctl==3.5.0
tzdata==2024.1
urllib3==2.2.2
wheel==0.44.0
xyzservices==2024.9.0
```

### How did you install OSMnx?

Pip

### Problem description

Most of the routing functions seem to handle undirected graphs very well. One exception is the `routing.route_to_gdf` function which occasionally breaks when using undirected graphs.

This is particularly a problem when using something like `routing.shortest_path` which can return a valid path on the undirected graph that `routing.route_to_gdf` cannot handle.

I'm not entirely sure what types of edges break `route_to_gdf` (maybe one-way streets?). The [current documentation for `routing.route_to_gdf`](https://osmnx.readthedocs.io/en/stable/user-reference.html#osmnx.routing.route_to_gdf) does expect a `MultiDiGraph`, however to maintain consistency with the other functions, this feels more like a bug than a feature request. Feel free to re-categorize as needed.

Please see the example below for a MWE. Thanks!

### Complete minimal reproducible example

```python
import networkx as nx
import osmnx as ox
import geopandas as gpd

location = [47.42185416567268, -92.93401175307781]
radius = 100
G = ox.graph_from_point(location, dist=radius)

UG = ox.convert.to_undirected(G)

path = ox.routing.shortest_path(UG, 1933044548, 1933008534)

# working    
ox.routing.route_to_gdf(G, path)

# broken
ox.routing.route_to_gdf(UG, path)
```

**Comment by gboeing** [comment]
There may be an easier solution here. I want to make sure this isn't an XY problem: can you tell me what your goal is for this model/analysis?

**Comment by gboeing** [comment]
The problem with allowing the `route_to_gdf` function to accept `MultiGraph` arguments is that it internally uses GeoPandas to identify the edges by `(u, v)` tuples (i.e., the order of `u` and `v` matters), of which there may be duplicates. However, an undirected graph's edges by definition are identified as `{u, v}` sets (i.e., the order of `u` and `v` doesn't matter), of which there may be duplicates. This is why you're receiving that error. Your path specifies a route from, say `c` to `b` to `a`, but undirected edge `{c, b}` is represented in the GeoDataFrame index as `(b, c)` so it cannot be found. There probably isn't a trivial fix for this.

OSMnx is designed to work with `MultiDiGraph` objects and is type-annotated and tested accordingly. Sometimes, passing a `MultiGraph` might work, but it isn't tested. It's just coincidental. Best to typecheck your code to avoid typing errors.

That said, how do you solve your problem? This is why I asked about the XY problem earlier. My best guess is that you want to allow bidirectional traffic along your network. Fortunately, OSMnx *is* designed to handle this using `MultiDiGraph` objects rather than `MultiGraph` objects. You just have to configure the `bidirectional_network_types` setting (see [the docs](https://osmnx.readthedocs.io/en/stable/user-reference.html#module-osmnx.settings)). Example:

```python
import osmnx as ox
ox.settings.bidirectional_network_types += "all"
G = ox.graph_from_point((47.42185, -92.93401), dist=100, network_type="all")
path = ox.routing.shortest_path(G, 1933044548, 1933008534)   
ox.routing.route_to_gdf(G, path)
```

Let me know if this addresses your goal. That said, the whole "use a bidirectional MultiDiGraph if you want an 'undirected' graph" concept isn't well-explained in the documentation, and may be worth explaining explicitly in the Getting Started guide.

**Comment by aarongraybill** [comment]
Thanks for getting back so fast. 

In terms of my underlying use case, I'm working on a project involving walking routes (trying to generate "optimal" routes for my running workouts)[^1]. The details aren't important, but the reason I'm thinking about `MultiGraphs` and not `MultiDiGraphs` is because there are no one-way streets on foot 😄. 

I'm also interested in generating routes that do not go on the same edge twice, so I was trying to explicitly collapse those same street, opposite direction edges. I had also assumed that fewer edges meant faster code.

It's not in my MWE, but I am using `G = ox.graph_from_point(location, dist=radius, network_type="walk", dist_type = "network")`, so I think I'm already taking care of the bidirectionality you suggested. I think my error was ever converting to a `MultiGraph` since I didn't understand that the bidirectionality was baked in.

To your final point, I had already read the Getting Started guide top-to-bottom before posting this, so I do think a bit more clarity there might be helpful (although I admit my use case was a bit specific). I would have personally benefitted from a bit more hand-holding on how walking differs from the other types of routing. I was basing my understanding off of this line in the [example notebook](https://github.com/gboeing/osmnx-examples/blob/main/notebooks/01-overview-osmnx.ipynb):

> 'walk' - get all streets and paths that pedestrians can use (this network type ignores one-way directionality)

I think you can go ahead and close this issue, though more documentation is always appreciated! Either way thanks for the excellent package, documentation, and interactive notebooks!

[^1]: Also don't worry, I'm not giving away my home address or anything. This is a deliberately obfuscated example.

**Comment by gboeing** [comment]
Thanks @aarongraybill, some thoughts:

> the reason I'm thinking about MultiGraphs and not MultiDiGraphs is because there are no one-way streets on foot

Yes, if you set `network_type="walk"` then OSMnx handles this accordingly by default (using MultiDiGraph objects). Otherwise, you just have to configure the `bidirectional_network_types` [setting](https://osmnx.readthedocs.io/en/stable/user-reference.html#module-osmnx.settings). Then for every edge from node `u` to node `v`, the graph will also have a reciprocal edge from `v` to `u` regardless of whether it's a one-way street or not. 

> I'm also interested in generating routes that do not go on the same edge twice, so I was trying to explicitly collapse those same street, opposite direction edges.

If you're solving a single shortest path, by definition it should not traverse the same edge twice, regardless. In graph theory, a path is defined as a sequence of adjacent edges in which all of the edges are distinct. If you're solving multiple shortest paths between a single origin/destination node pair, and you want the several paths to share no edges in common, then you should use a disjoint paths algorithm such as [this](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.connectivity.disjoint_paths.edge_disjoint_paths.html).

> I had also assumed that fewer edges meant faster code.

You're correct. Most graph algorithms' time complexity scales with the node count or edge count, sometimes quadratically. In practice, Dijkstra's algorithm and other path solvers have reasonable time complexity. If you're trying to solve many shortest paths (thousands or millions) and runtime becomes your bottleneck, you'll probably see greater gains by using OSMnx's parallelized path solving.

> I had already read the Getting Started guide top-to-bottom before posting this, so I do think a bit more clarity there might be helpful 

This is useful feedback. I think the resolution could be:
1. updating the getting started guide to point explicitly to the `settings` docs in the user guide so users know exactly what they can configure
2. updating the `convert.to_undirected` docstring to explain its limited use case: it really exists so you can get a MultiGraph object for those NetworkX graph algorithms which only accept a MultiGraph.
3. updating the `graph` module's `graph_from_x` functions' docstrings to explain that graph bidirectionality can be configured with the `settings` module rather than necessitating conversion to a MultiGraph

**Comment by gboeing** [comment]
Proposed fix in #1258.
